bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosBridgeHandler.java

```diff
@@ -0,0 +1,519 @@
+        logger.debug("Logging in to HEOS account.");
+        HeosResponseObject<Void> response = connection.logIn(username, password);
+
+        if (response.result) {
+            logger.debug("successfully logged-in, event is fired to handle post-login behaviour");
+            return;
+        }
+
+        @Nullable
+        HeosError error = response.getError();
+        logger.debug("Failed to login: {}", error);
+        updateStatus(ONLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                error != null ? error.code.toString() : "Failed to login, no error was returned.");
+    }
+
+    @Override
+    public void dispose() {
+        bridgeHandlerDisposalOngoing = true; // Flag to prevent the handler from being updated during disposal
+
+        cancel(startupFuture);
+        for (Future<?> future : childHandlerInitializedFutures) {
+            cancel(future);
+        }
+
+        @Nullable
+        HeosFacade localApiConnection = apiConnection;
+        if (localApiConnection == null) {
+            logger.debug("Not disposing bridge because of missing apiConnection");
+            return;
+        }
+
+        localApiConnection.unregisterForChangeEvents(this);
+        logger.debug("HEOS bridge removed from change notifications");
+
+        logger.debug("Dispose bridge '{}'", thing.getProperties().get(PROP_NAME));
+        localApiConnection.closeConnection();
+    }
+
+    /**
+     * Manages the removal of the player or group channels from the bridge.
+     */
+    @Override
+    public synchronized void childHandlerDisposed(ThingHandler childHandler, Thing childThing) {
+        logger.debug("Disposing child handler for: {}.", childThing.getUID().getId());
+        if (bridgeHandlerDisposalOngoing) { // Checks if bridgeHandler is going to disposed (by stopping the binding or
+            // openHAB for example) and prevents it from being updated which stops the
+            // disposal process.
+        } else if (childHandler instanceof HeosPlayerHandler) {
+            String channelIdentifier = "P" + childThing.getUID().getId();
+            updateThingChannels(channelManager.removeSingleChannel(channelIdentifier));
+        } else if (childHandler instanceof HeosGroupHandler) {
+            String channelIdentifier = "G" + childThing.getUID().getId();
+            updateThingChannels(channelManager.removeSingleChannel(channelIdentifier));
+            // removes the handler from the groupMemberMap that handler is no longer called
+            // if group is getting online
+            removeGroupHandlerInformation((HeosGroupHandler) childHandler);
+        }
+    }
+
+    @Override
+    public void childHandlerInitialized(ThingHandler childHandler, Thing childThing) {
+        logger.debug("Initialized child handler for: {}.", childThing.getUID().getId());
+        childHandlerInitializedFutures.add(scheduler.submit(() -> addPlayerChannel(childThing, null)));
```

Comment:
Should you also remove the future for a child when that child is disposed?

Reply:
It could but would make it more complex, now we might have one unplanned channel if an item is disposed again before its task was executed and it's fixed after the next restart. Very small chance and small impact. Can we keep it like this?

Reply:
I think the impact is almost negligible, so keep it as is.